### PR TITLE
esp_prov: switch from bluez/dbus to bleak (IDFGH-7556)

### PR DESCRIPTION
**Pull request by ccutrer**
to enable multiplatform BLE compatibility

tested on MacOS 12.4 and Ubuntu 20.04

also includes several discrete commits to allow interactive provisioning if necessary data isn't provided on command line.

### CODE REVIEW

**Comment by Alvin1Zhang** [comment]
Thanks for your contribution.

**Comment by mahavirj** [comment]
@ccutrer Thank you for sending this PR. It looks mighty useful, we are performing some tests on this and then we shall queue it for internal merge.

**Comment by HiFiPhile** [comment]
Tested working on Archlinux but failed on Windows 10 with error `Protocol Error 0x05: Insufficient Authentication`

Full log:
```
2022-06-13 16:28:57,669 bleak.backends.winrt.scanner DEBUG: Received 4A:5D:63:68:75:8F: Unknown.
2022-06-13 16:28:57,673 bleak.backends.winrt.scanner DEBUG: Received 27:E0:BB:A9:65:36: Unknown.
2022-06-13 16:28:57,739 bleak.backends.winrt.scanner DEBUG: 21 devices found. Watcher status: 3.
Connecting...
2022-06-13 16:28:57,778 bleak.backends.winrt.scanner DEBUG: Received 07:FE:CB:51:63:05: Unknown.
2022-06-13 16:28:57,783 bleak.backends.winrt.scanner DEBUG: Received 6E:97:48:68:7F:95: Unknown.
2022-06-13 16:28:57,784 bleak.backends.winrt.scanner DEBUG: Received 6E:97:48:68:7F:95: Unknown.
2022-06-13 16:28:57,896 bleak.backends.winrt.scanner DEBUG: Received 63:30:F1:98:32:00: Unknown.
2022-06-13 16:28:58,020 bleak.backends.winrt.scanner DEBUG: Received E7:5A:F5:6F:17:3D: Unknown.
2022-06-13 16:28:58,267 bleak.backends.winrt.scanner DEBUG: Received 62:B3:8E:1E:3D:0E: Unknown.
2022-06-13 16:28:58,268 bleak.backends.winrt.scanner DEBUG: Received 62:B3:8E:1E:3D:0E: Unknown.
2022-06-13 16:28:58,364 bleak.backends.winrt.scanner DEBUG: Received 7C:DF:A1:D1:D3:D6: Unknown.
2022-06-13 16:28:58,366 bleak.backends.winrt.scanner DEBUG: 6 devices found. Watcher status: 3.
2022-06-13 16:28:58,366 bleak.backends.winrt.client DEBUG: Connecting to BLE device @ 7C:DF:A1:D1:D3:D6
2022-06-13 16:29:01,156 bleak.backends.winrt.client DEBUG: session_status_changed_event_handler: id: <_bleak_winrt_Windows_Devices_Bluetooth.BluetoothDeviceId object at 0x0000000003BDFC50>, error: 0, status: 1
2022-06-13 16:29:01,156 bleak.backends.winrt.client DEBUG: Get Services...
2022-06-13 16:29:03,082 bleak.backends.winrt.client INFO: Services resolved for BleakClientWinRT (7C:DF:A1:D1:D3:D6)
Getting Services...
2022-06-13 16:29:03,202 bleak.backends.winrt.client DEBUG: Read Descriptor 0004 : bytearray(b'prov-scan')
2022-06-13 16:29:03,321 bleak.backends.winrt.client DEBUG: Read Descriptor 0007 : bytearray(b'prov-session')
2022-06-13 16:29:03,442 bleak.backends.winrt.client DEBUG: Read Descriptor 000A : bytearray(b'prov-config')
2022-06-13 16:29:03,562 bleak.backends.winrt.client DEBUG: Read Descriptor 000D : bytearray(b'proto-ver')
2022-06-13 16:29:03,682 bleak.backends.winrt.client DEBUG: Read Descriptor 0010 : bytearray(b'config')
Disconnecting...
Traceback (most recent call last):
  File "D:\mcu\esp-idf\tools\esp_prov\esp_prov.py", line 478, in <module>
    asyncio.run(main())
  File "C:\Program Files\Python39\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Program Files\Python39\lib\asyncio\base_events.py", line 642, in run_until_complete
    return future.result()
  File "D:\mcu\esp-idf\tools\esp_prov\esp_prov.py", line 374, in main
    if not await has_capability(obj_transport):
  File "D:\mcu\esp-idf\tools\esp_prov\esp_prov.py", line 112, in has_capability
    response = await tp.send_data('proto-ver', capability)
  File "D:\mcu\esp-idf\tools\esp_prov\transport\transport_ble.py", line 51, in send_data
    return await self.cli.send_data(self.name_uuid_lookup[ep_name], data)
  File "D:\mcu\esp-idf\tools\esp_prov\transport\ble_cli.py", line 153, in send_data
    await self.device.write_gatt_char(characteristic_uuid, bytearray(data.encode('latin-1')), True)
  File "C:\Tool\esp-idf\python_env\idf5.0_py3.9_env\lib\site-packages\bleak\backends\winrt\client.py", line 617, in write_gatt_char
    _ensure_success(
  File "C:\Tool\esp-idf\python_env\idf5.0_py3.9_env\lib\site-packages\bleak\backends\winrt\client.py", line 97, in _ensure_success
    raise BleakError(
bleak.exc.BleakError: Could not write value bytearray(b'none') to characteristic 000B: Protocol Error 0x05: Insufficient Authentication
```

**Comment by HiFiPhile** [comment]
Tested working on Archlinux but failed on Windows 10 with error `Protocol Error 0x05: Insufficient Authentication`

Full log:
```
2022-06-13 16:28:57,669 bleak.backends.winrt.scanner DEBUG: Received 4A:5D:63:68:75:8F: Unknown.
2022-06-13 16:28:57,673 bleak.backends.winrt.scanner DEBUG: Received 27:E0:BB:A9:65:36: Unknown.
2022-06-13 16:28:57,739 bleak.backends.winrt.scanner DEBUG: 21 devices found. Watcher status: 3.
Connecting...
2022-06-13 16:28:57,778 bleak.backends.winrt.scanner DEBUG: Received 07:FE:CB:51:63:05: Unknown.
2022-06-13 16:28:57,783 bleak.backends.winrt.scanner DEBUG: Received 6E:97:48:68:7F:95: Unknown.
2022-06-13 16:28:57,784 bleak.backends.winrt.scanner DEBUG: Received 6E:97:48:68:7F:95: Unknown.
2022-06-13 16:28:57,896 bleak.backends.winrt.scanner DEBUG: Received 63:30:F1:98:32:00: Unknown.
2022-06-13 16:28:58,020 bleak.backends.winrt.scanner DEBUG: Received E7:5A:F5:6F:17:3D: Unknown.
2022-06-13 16:28:58,267 bleak.backends.winrt.scanner DEBUG: Received 62:B3:8E:1E:3D:0E: Unknown.
2022-06-13 16:28:58,268 bleak.backends.winrt.scanner DEBUG: Received 62:B3:8E:1E:3D:0E: Unknown.
2022-06-13 16:28:58,364 bleak.backends.winrt.scanner DEBUG: Received 7C:DF:A1:D1:D3:D6: Unknown.
2022-06-13 16:28:58,366 bleak.backends.winrt.scanner DEBUG: 6 devices found. Watcher status: 3.
2022-06-13 16:28:58,366 bleak.backends.winrt.client DEBUG: Connecting to BLE device @ 7C:DF:A1:D1:D3:D6
2022-06-13 16:29:01,156 bleak.backends.winrt.client DEBUG: session_status_changed_event_handler: id: <_bleak_winrt_Windows_Devices_Bluetooth.BluetoothDeviceId object at 0x0000000003BDFC50>, error: 0, status: 1
2022-06-13 16:29:01,156 bleak.backends.winrt.client DEBUG: Get Services...
2022-06-13 16:29:03,082 bleak.backends.winrt.client INFO: Services resolved for BleakClientWinRT (7C:DF:A1:D1:D3:D6)
Getting Services...
2022-06-13 16:29:03,202 bleak.backends.winrt.client DEBUG: Read Descriptor 0004 : bytearray(b'prov-scan')
2022-06-13 16:29:03,321 bleak.backends.winrt.client DEBUG: Read Descriptor 0007 : bytearray(b'prov-session')
2022-06-13 16:29:03,442 bleak.backends.winrt.client DEBUG: Read Descriptor 000A : bytearray(b'prov-config')
2022-06-13 16:29:03,562 bleak.backends.winrt.client DEBUG: Read Descriptor 000D : bytearray(b'proto-ver')
2022-06-13 16:29:03,682 bleak.backends.winrt.client DEBUG: Read Descriptor 0010 : bytearray(b'config')
Disconnecting...
Traceback (most recent call last):
  File "D:\mcu\esp-idf\tools\esp_prov\esp_prov.py", line 478, in <module>
    asyncio.run(main())
  File "C:\Program Files\Python39\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Program Files\Python39\lib\asyncio\base_events.py", line 642, in run_until_complete
    return future.result()
  File "D:\mcu\esp-idf\tools\esp_prov\esp_prov.py", line 374, in main
    if not await has_capability(obj_transport):
  File "D:\mcu\esp-idf\tools\esp_prov\esp_prov.py", line 112, in has_capability
    response = await tp.send_data('proto-ver', capability)
  File "D:\mcu\esp-idf\tools\esp_prov\transport\transport_ble.py", line 51, in send_data
    return await self.cli.send_data(self.name_uuid_lookup[ep_name], data)
  File "D:\mcu\esp-idf\tools\esp_prov\transport\ble_cli.py", line 153, in send_data
    await self.device.write_gatt_char(characteristic_uuid, bytearray(data.encode('latin-1')), True)
  File "C:\Tool\esp-idf\python_env\idf5.0_py3.9_env\lib\site-packages\bleak\backends\winrt\client.py", line 617, in write_gatt_char
    _ensure_success(
  File "C:\Tool\esp-idf\python_env\idf5.0_py3.9_env\lib\site-packages\bleak\backends\winrt\client.py", line 97, in _ensure_success
    raise BleakError(
bleak.exc.BleakError: Could not write value bytearray(b'none') to characteristic 000B: Protocol Error 0x05: Insufficient Authentication
```

File: tools/esp_prov/transport/ble_cli.py [hunk 1]
```diff
@@ -46,23 +50,56 @@ async def connect(self, devname, iface, chrc_names, fallback_srv_uuid):
...
+                print('No devices found!')
+                exit(1)
+
+            while True:
+                devices.sort(key=device_sort)
+                print('==== BLE Discovery results ====')
+                print('{0: >4} {1: <33} {2: <12}'.format(
+                    'S.N.', 'Name', 'Address'))
+                for i in range(len(devices)):
+                    print('[{0: >2}] {1: <33} {2: <12}'.format(i + 1, devices[i].name or 'Unknown', devices[i].address))
```
[the remark this answers is not in the document]
Comment (ccutrer) [reply tools/esp_prov/transport/ble_cli.py:66]: are you sure you're running the latest commit? I pushed up a fix for `unsupported format string passed to NoneType.__format__` (adding `or 'Unknown'`) right around the time you commented that it's working on Windows 10. Assuming you are, would the fix be to change it to `str(device[I].name or 'Unknown')`? I'm a relative novice in Python.

[the remark this answers is not in the document]
Comment (HiFiPhile) [reply tools/esp_prov/transport/ble_cli.py:66]: > are you sure you're running the latest commit? 

You are right, it's already fixed in last commit.

**Comment by laukik-hase** [comment]
Hello, @ccutrer!

Thank you for your contribution! Could you please rebase your PR with the latest `master` and resolve the merge conflicts?

**Comment by ccutrer** [comment]
@laukik-hase: done.

**Comment by laukik-hase** [comment]
sha=265ad3eea657fb3d4f064fbb49132ff8b128d1b8

**Comment by laukik-hase** [comment]
Merged with https://github.com/espressif/esp-idf/commit/3f77c65e56bfc2ce8255af1fe68ce6b056bc993a

**Pull request closed without merging**